### PR TITLE
Fix handling of "leader changed" errors

### PR DESCRIPTION
**Pull request by cenkalti**
**What this PR does / why we need it**:
This PR aims to fix temporary "etcdserver: leader changed" errors from kube-apiserver that is previously attempted by #11401 by adding a single retry when this kind of error is detected.

/cc @dims 

**Special notes for your reviewer**:
I also reverted the previous fix that didn't solve the issue. One commit is the revert, the other one is the new fix. It's better if you review commits separately.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by dims** [comment]
@hickeyma @technosophos @mattfarina Folks, @cenkalti tested a bunch of scenarios recreating the problem(s) with etcd with this patch.

File: pkg/cli/roundtripper.go
Comment (dims) [inline pkg/cli/roundtripper.go:46]: @cenkalti is there anything the http headers that would indicate that we are dealing with an error? (so we can limit reading the full body only to the error scenarios?)

Comment (cenkalti) [reply pkg/cli/roundtripper.go:46]: Unfortunately no. The headers sent by kube-apiserver are as below:
```
HEADER Cache-Control no-cache, private
HEADER Content-Type application/json
HEADER X-Kubernetes-Pf-Flowschema-Uid 731b79e0-c1d1-4d8d-bba8-7588c969de79
HEADER X-Kubernetes-Pf-Prioritylevel-Uid 251f7bf0-8d47-4e8e-b644-6077a1fe19ab
HEADER Content-Length 119
HEADER Date Tue, 11 Oct 2022 13:17:31 GMT
HEADER Audit-Id 15493c2a-d08a-4410-b634-7b1302707bf5
```

**Comment by cenkalti** [comment]
Hey @technosophos @hickeyma. Bumping up this for visibility. It's a small fix. Can you take a look please?

**Review by hickeyma — COMMENTED** [review]
@cenkalti thanks for pushing the PR. The code looks good in principle. However, we don't have any easy way of testing this and this is the 2nd fix following #11401. Would it be possible to provide some tests for this?

**Comment by cenkalti** [comment]
@hickeyma I tested this fix manually by setting up a custom k8s cluster with 3 etcd nodes and running `etcdctl move-leader` with random member ID continuously to trigger the error case. Even with this special setup, the error does not occur every time when I run `helm install` command as the timing of API calls are very critical that they must overlap with the leader change event.

I checked https://github.com/helm/acceptance-testing project to see If I can write this scenario as an acceptance test but I couldn't find a solution as the clusters are created with `kind` and there is no easy way to trigger a leader change event.

**Review by hickeyma — APPROVED** [review]
LGTM, thanks @cenkalti

**Comment by hickeyma** [comment]
Needs another maintainer approval for merge.

**Review by dims — APPROVED** [review]
LGTM

**Comment by dims** [comment]
thanks @cenkalti @hickeyma

**Comment by cenkalti** [comment]
@technosophos @mattfarina Can you take a look please?

**Review by technosophos — APPROVED** [review]
Looks good.

**Comment by technosophos** [comment]
@hickeyma feel free to merge whenever you feel good about it.

**Comment by sruthiwander** [comment]
Hi, Any chance this could be patched to the 3.2.x release.

**Pull request merged**
